### ISSUE TITLE
Generic NES-like controller always holds down left, and --gamepad_override doesn't seem to fix it

### ISSUE THREAD

**Issue by JacobKfromIRC** [divVerent/aaaaxy#28]
I have a generic NES-like controller branded "retro power" that the game interprets as always holding down left. The controller works fine in AntiMicroX and Taisei, but for some reason AAAAXY seems to misread the controller. I tried running `./AAAAXY-x86_64.AppImage --gamepad_override "0300000079000000110000001001000012117,Retrolink SNES Controller,platform:Linux,a:b2,b:b1,x:b3,y:b0,back:b8,start:b9,leftshoulder:b4,rightshoulder:b5,dpup:a1,dpleft:a0,dpdown:a1,dpright:a0," (I got the gamepad override string from AntiMicroX. The gamepad doesn't really have that many buttons.), but that seemed to have no effect. Actually, I just now noticed it says "[INFO] gamepad mappings from --gamepad_override applied" regardless of whether I actually use the "--gamepad_override" option.

I have tried "AAAAXY 1.2.280+20220822.2294.9ff4b8b7" and "AAAAXY 1.2.321+20220905.2335.e05b1efe", both via the AppImage, and I am using Trisquel GNU/Linux Nabia (10.0.1) as my operating system. Unfortunately, I could not find a controller that actually works with AAAAXY. I have another controller that came with the NES-like one, and it has a similar problem, and I have a couple of Wii remotes, which do not have the holding-down-left problem, but the D-pad is rotated and none of the buttons jump or pause. The Wii remote D-pad controls the entire operating system for some reason though, so I think the D-pad rotation is a operating system problem or something (Actually, the way the D-pad is rotated would make sense if I held the Wii remote like a TV remote.).

**Comment by divVerent** [comment]
First of all, the mappings are, like all non-debug flags, stored in ~/.config/AAAAXY/config.json - so that not specifying the flag still shows that they're used is fine.

My first suspicion is that the override isn't actually applied to this gamepad; I'll add debug logging to show the actual SDL ID.

Once your mapping is applied, you'll probably be able to locate the bad assignment easily. Having both dpleft and dpright assigned to a0 seems a bit suspicious, BTW; normally one is assigned to -a0 and the other to +a0.

**Comment by divVerent** [comment]
Next release will give you the SDLID of the gamepad that the game library uses - this should help with overriding and fixing the mappings.

**Comment by divVerent** [comment]
That release is out - first thing to check is which ID your controller gets. It is possible that the wrong ID is used and that is why your remapping failed.

**Comment by JacobKfromIRC** [comment]
Yep, the ID was incorrect. AAAAXY shows it as 03000000790000001100000010010000, and when I change the ID in gamepad_override to that, it no longer acts like I'm always holding down left (but the D-pad doesn't do anything). (I wonder why AntiMicroX shows what it does?).

I tried `./AAAAXY-x86_64.AppImage --gamepad_override "03000000790000001100000010010000,Retrolink SNES Controller,platform:Linux,a:b2,b:b1,back:b8,start:b9,dpup:-a1,dpdown:+a1,dpleft:-a0,dpright:+a0"` and then I get the problem with always holding left again.

In jstest-gtk the controller seems fine.

**Comment by divVerent** [comment]
I wonder if something is odd with how this controller reports axis values.
Fixes to github.com/hajimehoshi/ebiten may be necessary for this.

Last four digits of the ID looks suspicious. Ebitengine comes with an
example program for gamepads - what axis 0 values does it show for left,
nothing pressed and right?

**Comment by divVerent** [comment]
So... first of all, we can always use gamepad definitions to turn unwanted axes off.

As for how the game handles multiple inputs: if _any_ left key is pressed, the player moves left, unless there is also _any_ right key pressed. I might change this to differencing logic instead "more left keys than right keys", 

The concrete gamepad appears weird indeed, however it seems only the d-pad data that is broken, while the axes are fine?

> Another thing I noticed is the "Axes:" value at the top. When I'm not holding anything on the gamepad, it shows "0:-0.00, 1:-0.00" which doesn't seem too weird at first, and interacting with the controller works as you'd expect, with holding left resulting in "0:-1.00, 1:-0.00", holding up resulting in "0:-0.00, 1:-1.00", holding right resulting in "0:+1.00, 1:-0.00", and holding down resulting in "0:-0.00, 1:+1.00".

This actually looks good; I see good chances that this is something that can be made usable somehow.

Try this string then:

```
03000000790000001100000010010000,Retrolink SNES Controller,platform:Linux,a:b2,b:b1,back:b8,start:b9,leftx:a0,lefty:a1
```

This isn't _quite_ correct as it treats the D-pad as an analog stick, but for AAAAXY that will work fine.

As for A/B button mapping, this game should play equally well either way; I intended to have jumping on A as that feels more natural on an Xbox-style pad (although one can also play with the X and Y buttons, which have the same but reversed), but on SMB it was exactly the other way round. I guess it's your preference then :)

**Comment by divVerent** [comment]
I found in Ebitengine source code that the function `buttonValue` only maps the range [-1, 1] to buttons (like dpad); so normally -1 is treated as "not pressed" and +1 is treated as "pressed". The threshold appears to be -0.76. However if you write +a0, Ebitengine maps [0, 1] to [-1, 1] first, making the threshold 0.12, and if you write -a0, Ebitengine maps [-1, 0] to [-1, 1] first. This appears wrong to me, but nevertheless, you might be able to fix the dpad mappings by inverting (using -a0~ instead of -a0). If that works, might be worth reporting a bug to Ebitengine.

**Comment by divVerent** [comment]
Yes, this seems wrong in Ebitengine (after comparing to SDL source), I'll file a bug.

**Comment by divVerent** [comment]
Also, can you try if https://github.com/hajimehoshi/ebiten/pull/2335 fixes everything for you in the gamepad demo of Ebitengine? There was indeed a nasty bug affecting all negative axis mappings.

**Comment by divVerent** [comment]
I released with this fix. Enjoy!

**Comment by JacobKfromIRC** [comment]
I tried the leftx/lefty mapping and the -a0~ mapping with AAAAXY 1.2.331+20220912.2345.cbfba5a3 and those both worked fine, and I then deleted config.json to clear the mappings and updated to the latest version (AAAAXY 1.2.336+20220919.2350.a525d576), and the controller still worked fine. I haven't gotten around to testing with the updated ebitengine gamepad test (though, not sure if that makes sense to do given it's fixed in AAAAXY which incorporates ebiten).

I guess this issue can be closed now, as the controller works fine by default now.

Currently, the default is that physical A is jump. I tried using physical B for jump as a test and it felt really strange, but that could just be because I've played a lot of Super Mario Bros., so I'm used to holding down B and pressing A. I think the current mapping should stay. I think Super Mario World (which switched to a 4-face-button controller) started using B to jump, with B at the bottom (where A is on the Xbox pad), so A to jump matches Nintendo for NES and Xbox controllers. But then, A and B are swapped compared to Xbox on modern Nintendo controllers, so it might make more sense to define jump as the lower button rather than a letter, if that's possible.
